daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala

```diff
@@ -0,0 +1,272 @@
+
+    // The List is necessary because the rules can:
+    // (1) drop nodes; (2) combine nodes (3) lift nodes from a lower level to a higher level.
+
+    txOriginal match {
+      case GenTransaction(nodesOriginal, rootsOriginal) =>
+        def traverseNids[R](xs: List[Nid])(k: List[Norm] => Tramp[R]): Tramp[R] = {
+          Bounce { () =>
+            xs match {
+              case Nil => k(Nil)
+              case x :: xs =>
+                traverseNode(nodesOriginal(x)) { norms1 =>
+                  traverseNids(xs) { norms2 =>
+                    Bounce { () =>
+                      k(norms1 ++ norms2)
+                    }
+                  }
+                }
+            }
+          }
+        }
+
+        def traverseNode[R](node: Node)(k: List[Norm] => Tramp[R]): Tramp[R] = {
+          Bounce { () =>
+            node match {
+
+              case NodeRollback(children) =>
+                traverseNids(children.toList) { norms =>
+                  makeRoll(norms)(k)
+                }
+
+              case exe: NodeExercises[_, _] =>
+                traverseNids(exe.children.toList) { norms =>
+                  k(List(Norm.Exe(exe, norms)))
+                }
+
+              case leaf: LeafOnlyActionNode[_] =>
+                k(List(Norm.Leaf(leaf)))
+            }
+          }
+        }
+        //pass 1
+        traverseNids(rootsOriginal.toList) { norms =>
+          //pass 2
+          pushNorms(initialState, norms) { (finalState, roots) =>
+            Land(GenTransaction(finalState.nodeMap, ImmArray(roots)))
+          }
+        }.run
+    }
+  }
+
+  // makeRoll: encodes the normalization transformation rules:
+  //   rule #1: ROLL [ ] -> ε
+  //   rule #2: ROLL [ ROLL [ xs… ] , ys… ] -> ROLL [ xs… ] , ROLL [ ys… ]
+  //   rule #3: ROLL [ xs… , ROLL [ ys… ] ] ->  ROLL [ xs… , ys… ]
+
+  //   rule #2/#3 overlap: ROLL [ ROLL [ xs… ] ] -> ROLL [ xs… ]
+
+  private def makeRoll[R](norms: List[Norm])(k: List[Norm] => Tramp[R]): Tramp[R] = {
+    caseNorms(norms) match {
+      case Case.Empty =>
+        // normalization rule #1
+        k(Nil)
+
+      case Case.Single(roll: Norm.Roll) =>
+        // normalization rule #2/#3 overlap
+        k(List(roll))
+
+      case Case.Single(act: Norm.Act) =>
+        // no rule
+        k(List(Norm.Roll1(act)))
+
+      case Case.Multi(h: Norm.Roll, m, t) =>
+        // normalization rule #2
+        makeRoll(m ++ List(t)) { norms =>
+          k(h :: norms)
+        }
+
+      case Case.Multi(h: Norm.Act, m, t: Norm.Roll) =>
+        // normalization rule #3
+        k(List(pushIntoRoll(h, m, t)))
+
+      case Case.Multi(h: Norm.Act, m, t: Norm.Act) =>
+        // no rule
+        k(List(Norm.Roll2(h, m, t)))
+    }
+  }
+
+  private def pushIntoRoll(a1: Norm.Act, xs2: List[Norm], t: Norm.Roll): Norm.Roll = {
+    t match {
+      case Norm.Roll1(a3) => Norm.Roll2(a1, xs2, a3)
+      case Norm.Roll2(a3, xs4, a5) => Norm.Roll2(a1, xs2 ++ List(a3) ++ xs4, a5)
```

Comment:
```suggestion
      case Norm.Roll2(a3, xs4, a5) => Norm.Roll2(a1, xs2 ++ a3 +: xs4, a5)
```
Anyway is concatenating lists a good idea ? 
Either add a comment to explain why this is not linear, either you use better data structure (e.g. Frontstack, Backstack, Vector ...)